### PR TITLE
Add [Dwarf_high] library

### PR DESCRIPTION
**Pull request by mshinwell**
The library added by this pull request contains various functions and abstractions to assist with the generation of DWARF information.  It is based on #2077.

Some of this code has been rewritten this week and is awaiting testing, which will be done shortly.  However I don't expect much change from this to the final state.

@bschommer  Please look at this too if you could.  It should put the `Dwarf_low` library into context somewhat.

### CODE REVIEW

**Comment by mshinwell** [comment]
(There are a couple of changes to integer-related modules in this patch which can be split out into a separate pull request once testing has been completed.)

**Comment by bschommer** [comment]
I will have look at this in the next few days.

File: asmcomp/debug/dwarf_high/operator_builder.ml [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+  (* Same comment as per [address_of_stack_slot]. *)
+  [ O.DW_op_call_frame_cfa;
+    O.DW_op_consts offset_in_bytes;
+    O.DW_op_minus;
+    O.DW_op_deref;
+  ]
+
+let value_of_symbol ~symbol : O.t = DW_op_addr symbol
+
+let signed_int_const i : O.t = DW_op_consts i
```
Comment (bschommer) [inline asmcomp/debug/dwarf_high/operator_builder.ml]: Given that there are now the different size variant of DW_op_const maybe you could chose the one that fits best to safe some space.

File: asmcomp/debug/dwarf_high/operator_builder.ml [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+let add_unsigned_const i : O.t =
+  if Targetint.compare i Targetint.zero < 0 then begin
+    Misc.fatal_error "[Operator_builder.add_unsigned_const] only takes \
+      integers >= 0"
+  end;
+  DW_op_plus_uconst i
```
Comment (bschommer) [inline asmcomp/debug/dwarf_high/operator_builder.ml]: Same comment as above

**Comment by bschommer** [comment]
Looks good overall and I could not find any problems with the code beside the two comments.

**Comment by mshinwell** [comment]
@bschommer Thanks, will look at comments.  Let's pause review now until I've finished doing some changes in `dwarf.ml` and related files -- I'll ping again when those have been done.

**Comment by mshinwell** [comment]
@bschommer There were various bugs, misfeatures and missing features in this library which have now been fixed.  Please see `dd34512` for the latest version.  Note that this branch does not yet include the latest version of the `dwarf_low/` library.

**Pull request closed without merging**